pkg/controller/shared_state.go

```diff
@@ -1135,20 +1135,37 @@ func (c *SharedState) cleanUpAzVolumeAttachmentByNode(ctx context.Context, azDri
 func (c *SharedState) cleanUpAzVolumeAttachments(ctx context.Context, attachments []azdiskv1beta2.AzVolumeAttachment, cleanUp attachmentCleanUpMode, caller operationRequester) error {
 	var err error
 
+	var markDetachRequest = func(attachment *azdiskv1beta2.AzVolumeAttachment, patchRequired *bool) {
+		attachment.Status.Annotations = azureutils.AddToMap(attachment.Status.Annotations, consts.VolumeDetachRequestAnnotation, string(caller))
+		*patchRequired = true
+	}
+
+	var markCleanUp = func(attachment *azdiskv1beta2.AzVolumeAttachment, patchRequired *bool) {
+		attachment.Status.Annotations = azureutils.AddToMap(attachment.Status.Annotations, consts.CleanUpAnnotation, string(caller))
+		*patchRequired = true
+	}
+
 	for _, attachment := range attachments {
 		var patchRequired bool
 		patched := attachment.DeepCopy()
 
-		// if caller is azdrivernode, don't append cleanup annotation
-		if (caller != azdrivernode && !metav1.HasAnnotation(patched.ObjectMeta, consts.CleanUpAnnotation)) ||
-			// replica attachments should always be detached regardless of the cleanup mode
-			((cleanUp == cleanUpAttachment || patched.Spec.RequestedRole == azdiskv1beta2.ReplicaRole) && !metav1.HasAnnotation(patched.ObjectMeta, consts.VolumeDetachRequestAnnotation)) {
-			patchRequired = true
-			if caller != azdrivernode {
-				patched.Status.Annotations = azureutils.AddToMap(patched.Status.Annotations, consts.CleanUpAnnotation, string(caller))
+		if attachment.Spec.RequestedRole == azdiskv1beta2.PrimaryRole {
+			if cleanUp == cleanUpAttachment && !volumeDetachRequested(patched) {
+				markDetachRequest(patched, &patchRequired)
+			} else if deleteRequested, _ := objectDeletionRequested(&attachment); !deleteRequested {
+				// if primary azvolumeattachments are being cleaned up for driver uninstall, issue a DELETE call and continue
+				if err = c.cachedClient.Delete(ctx, patched); err != nil {
+					return err
+				}
+			}
+		} else {
+			// replica mount should always be detached in cleanup regardless to the cleanup mode
+			if !volumeDetachRequested(patched) {
+				markDetachRequest(patched, &patchRequired)
```

Comment:
If we markDetachRequest for replica AzVolumeAttachments, we couldn't recover them from VMs because them are detached already when we re-install driver. Then this PR will make no sense: https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1666

Sould we do the same for both Replica and Primary roles here(only delete them by calling: c.cachedClient.Delete())? Don't detach any disks while uninstalling driver.

Reply:
Replica `AzVolumeAttachments` are resources only used in v2 driver. That said, if we don't detach replica `AzVolumeAttachment` on v2 driver uninstallation, the user will end up with disk attachments unknown to k8s.

The main objective of https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1666 is so that we can properly recover `AzVolumeAttachment.Status.State` on driver reinstall. And being able to recreate replica `AzVolumeAttachment` that had failed to detach for any reason during driver uninstall is an additional benefit this PR will bring.

Reply:
I see! Thanks for the information.